crypto/evp_extra/p_ed25519_asn1.c

```diff
@@ -43,10 +48,17 @@ static int ed25519_set_priv_raw(EVP_PKEY *pkey, const uint8_t *in, size_t len) {
 
   // The RFC 8032 encoding stores only the 32-byte seed, so we must recover the
   // full representation which we use from it.
-  uint8_t pubkey_unused[32];
-  ED25519_keypair_from_seed(pubkey_unused, key->key, in);
+  uint8_t pubkey_computed[32];
+  ED25519_keypair_from_seed(pubkey_computed, key->key, privkey);
   key->has_private = 1;
 
+  // If a public key was provided, validate that it matches the computed value.
+  if(pubkey && OPENSSL_memcmp(pubkey_computed, pubkey, pubkey_len) != 0) {
```

Comment:
The current ed25519 logic already computes the public key, so this isn't adding any additional scalar multiplications. This is just merely validating that there is consistency between the public key that was encoded in the file, and the one that was computed from the private key.  This was a behavior Ring was doing as well, but is not driven by any requirement from any RFC that I've reviewed.

This is essentially protecting that there can't be a mismatched private and public key pair present within the PKCS#8 v2 structure. A good data point to look at is how we handle [ECPrivateKey](https://www.rfc-editor.org/rfc/rfc5915.html) for elliptic curve keys. Today we only compute the public key if it wasn't embedded within the private key structure, otherwise it will use the public key in the encoding. Maybe we update ed25519 to do the same?

It's still an interesting question Ring has proposed on whether we should validate that the pair matches rather then trust by default.

Reply:
I believe I recall that validating curve25519 public keys was debated, but I don't recall the conclusion of those public discussions. Need to do some research.

> Maybe we update ed25519 to do the same?

What I meant was that this case, when the public key is embedded in the v2 encoding, then we save computations.

Reply:
If the signature algorithm is ever invoked with two different key pairs `(sk, pk1)` and `(sk, pk2)`, where `pk1 != pk2`, then we have an issue; you can recover `sk`, cf. https://crypto.stackexchange.com/a/13141.

Because of this, in the face of our API maybe theoretical, I propose to keep this check. This is also considering that this code-path is likely not the hottest - but it does add latency to parsing the key.

Can we add this to the benchmarking tool?